mlir/lib/Conversion/ConvertToAIRPass.cpp

```diff
@@ -2389,6 +2650,87 @@ struct CopyToDmaPass : public air::impl::CopyToDmaBase<CopyToDmaPass> {
 
     LLVM_DEBUG(llvm::outs() << "output\n");
     LLVM_DEBUG(module.print(llvm::outs()));
+
+    // Condense memref data pattern reordering ops, including memref.subview,
+    // memref.tranpose and memref.expand_shape into air.dma_memcpy_nd op's
+    // offsets, sizes and strides fields.
+    auto scope = getOperation();
+    std::vector<std::tuple<xilinx::air::DmaMemcpyNdOp, std::vector<Operation *>,
+                           std::vector<Operation *>>>
+        dma_ops;
+
+    scope->walk([&](xilinx::air::DmaMemcpyNdOp dmaOp) {
+      bool src_condense = false;
+      if (auto src_defop = dmaOp.getSrcMemref().getDefiningOp()) {
+        src_condense |= isa<memref::TransposeOp>(src_defop);
+        src_condense |= isa<memref::ExpandShapeOp>(src_defop);
+        src_condense |= isa<memref::SubViewOp>(src_defop);
+      }
+      bool dst_condense = false;
+      if (auto dst_defop = dmaOp.getDstMemref().getDefiningOp()) {
+        dst_condense |= isa<memref::TransposeOp>(dst_defop);
+        dst_condense |= isa<memref::ExpandShapeOp>(dst_defop);
+        dst_condense |= isa<memref::SubViewOp>(dst_defop);
+      }
+      if (src_condense || dst_condense) {
+        std::tuple<air::DmaMemcpyNdOp, std::vector<Operation *>,
+                   std::vector<Operation *>>
+            log_entry;
+        std::get<0>(log_entry) = dmaOp;
+        if (src_condense) {
+          Operation *ancestor = dmaOp.getSrcMemref().getDefiningOp();
+          bool exit = false;
+          while (ancestor && !exit) {
+            if (auto transpose_anc = dyn_cast<memref::TransposeOp>(ancestor)) {
+              std::get<1>(log_entry).push_back(ancestor);
+              ancestor = transpose_anc.getIn().getDefiningOp();
```

Comment:
What if ancestor has more than one consumer -- just thinking ahead here, in case the ancestor is removed, which seems to be what happens at the end of this function.

Reply:
This is a great point. I changed it to not delete the chain of memref ops, and instead rely on the `-canonicalize` to remove the ops for me. Thanks for pointing it out.